launcher/internal/logging/logging.go

```diff
@@ -0,0 +1,255 @@
+	serialConsoleFile *os.File
+}
+
+type payload map[string]any
+
+// NewLogger returns a Logger with Cloud and Serial Console logging configured.
+func NewLogger(ctx context.Context) (Logger, error) {
+	// Retrieve monitored resource information.
+	mdsClient := metadata.NewClient(nil)
+
+	projectID, err := mdsClient.ProjectIDWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceID, err := mdsClient.InstanceIDWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceName, err := mdsClient.InstanceNameWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	zone, err := mdsClient.ZoneWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Configure Cloud Logging client/logger.
+	cloggingClient, err := clogging.NewClient(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Configure Serial Console logger.
+	serialConsole, err := os.OpenFile(serialConsoleFile, os.O_WRONLY, 0)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open serial console for writing: %v", err)
+	}
+
+	slg := slog.New(slog.NewTextHandler(serialConsole, nil))
+	slg.Info("Serial Console logger initialized")
+
+	// This is necessary for DEBUG logs to propagate properly.
+	slog.SetDefault(slg)
+
+	return &logger{
+		cloudLogger:  cloggingClient.Logger(logName),
+		serialLogger: slg,
+		resource: &mrpb.MonitoredResource{
+			Type: "gce_instance",
+			Labels: map[string]string{
+				"project_id":  projectID,
+				"instance_id": instanceID,
+				"zone":        zone,
+			},
+		},
+		instanceName:      instanceName,
+		cloudClient:       cloggingClient,
+		serialConsoleFile: serialConsole,
+	}, err
+}
+
+func (l *logger) SerialConsoleFile() *os.File {
+	return l.serialConsoleFile
+}
+
+func (l *logger) Close() {
+	if l.cloudClient != nil {
+		l.cloudClient.Close()
+	}
+
+	if l.serialConsoleFile != nil {
+		l.serialConsoleFile.Close()
+	}
+}
+
+// Given a list of args, recursively converts it to a payload.
+// Assumes alternating keys and values (mirroring slog's behavior).
+func addArgs(pl payload, args []any) {
+	// Base case - if args is empty.
+	if len(args) == 0 {
+		return
+	}
+
+	// Base case - if args has one element.
+	if len(args) == 1 {
+		// If the arg is a valid key, add with empty value.
+		key, ok := args[0].(string)
+		if ok {
+			pl[key] = ""
+		}
+		return
+	}
+
+	key, ok := args[0].(string)
+	if ok {
+		// If key is a valid string, add pair to payload. Otherwise, the pair is skipped.
+		pl[key] = args[1]
+	}
+
+	// Recurse with remaining args.
+	addArgs(pl, args[2:])
+}
+
+func (l *logger) writeLog(severity clogging.Severity, msg string, args ...any) {
+	// Write cloud log.
+	logEntry := clogging.Entry{
+		Severity: severity,
+		Resource: l.resource,
+	}
+
+	pl := payload{}
+	addArgs(pl, args)
+
+	if len(msg) > 0 {
+		pl[payloadMessageKey] = msg
+	}
+
+	if len(l.instanceName) > 0 {
+		// Needed for backwards compatibility with Cloudbuild tests.
+		pl[payloadHostnameKey] = l.instanceName
```

Comment:
You can feel free to change the tests. That was just the quickest way I had to write that script

Reply:
The field is still worth logging for filtering purposes but changed the key.

Reply:
Scratch that - I haven't been able to get workload log redirection to Cloud Logging working. Going to make that future work and leave the keys as-is for test backcompat.